src/Orleans/Streams/Internal/StreamSubscriptionChangeHandler.cs

```diff
@@ -0,0 +1,91 @@
+        /// </summary>
+        /// <param name="subscriptionId"></param>
+        /// <param name="streamId"></param>
+        /// <param name="messageType"></param>
+        /// <returns></returns>
+        public async Task HandleNewSubscription(GuidId subscriptionId, StreamId streamId, Type messageType)
+        {
+            // if no observer attached to the subscription, check if the grain is a IStreamSubscriptionObserver
+            IStreamSubscriptionObserverProxy observerProxy;
+            if (TryGetStreamSubscriptionObserverProxyForType(messageType, out observerProxy))
+            {
+                var streamProvider = this.providerRuntime.ServiceProvider
+                            .GetService<IStreamProviderManager>()
+                            .GetStreamProvider(streamId.ProviderName);
+
+                await observerProxy.InvokeOnNewSubscription(streamId, subscriptionId, streamProvider);
+            }
+        }
+
+        private bool TryGetStreamSubscriptionObserverProxyForType(Type concretType, out IStreamSubscriptionObserverProxy observerProxy)
+        {
+            IStreamSubscriptionObserverProxy proxy = null;
+            foreach (var observerEntry in this.subscriptionObserverMap)
+            {
+                if (observerEntry.Key.IsAssignableFrom(concretType))
+                {
+                    proxy = observerEntry.Value;
+                    break;
+                }
+            }
+            observerProxy = proxy;
+            return proxy != null;
+        }
+    }
+
+    internal interface IStreamSubscriptionObserverProxy 
+    {
+        Task InvokeOnNewSubscription(StreamId streamId, GuidId subscriptionId, IStreamProvider streamProvider);
+        IAddressable SubscriptionObserver { set; }
+    }
+
+    /// <summary>
+    /// Decorator class for a IStreamSubscriptionObserver<T>. Created mainly to avoid using reflection to invoke
+    /// methods on IStreamSubscriptionObserver<T>.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal class StreamSubscriptionObserverProxy<T> : IStreamSubscriptionObserver<T>, IStreamSubscriptionObserverProxy
+    {
+        private IStreamSubscriptionObserver<T> observer => this.SubscriptionObserver as IStreamSubscriptionObserver<T>;
+        public IAddressable SubscriptionObserver { private get; set; }
+        public StreamSubscriptionObserverProxy()
+        {
+        }
+
+        public Task OnNewSubscription(StreamSubscriptionHandle<T> handle)
+        {
+            return this.observer.OnNewSubscription(handle);
+        }
+
+        public Task InvokeOnNewSubscription(StreamId streamId, GuidId subscriptionId, IStreamProvider streamProvider)
+        {
+            var stream = streamProvider.GetStream<T>(streamId.Guid, streamId.Namespace) as StreamImpl<T>;
+            var handle = new StreamSubscriptionHandleImpl<T>(subscriptionId, stream);
+            return this.observer.OnNewSubscription(handle);
```

Comment:
should probably call to classes own OnNewSubscription which calls the observer's.  That way the observer's is called in only one place and if we need to add logging or statistics we only have one entry point.